### PR TITLE
Fix #6046: Wallet show solana account creation via dapp

### LINKED ISSUE #6046: Implement show account creation flow
### Description: 

As part of Solana dapps work, we have a new `BraveWalletProviderDelegate` function `func showAccountCreation(_ type: BraveWallet.CoinType)` which is called when a user tries to connect to a Solana dapp without a Solana account/keyring. As of 1.44, users will have a Solana account by default however users who setup their Wallet before 1.44 will not have one and we need to show the account creation flow when this delegate function is called.

### PR DESCRIPTION
**Pull request by nuo-xu**
## Summary of Changes
solana provider delegate call back will trigger a solana account creation flow once the dapp is trying to connect to a brave wallet that does not have a solana account.
Once user has created the new solana account (which will be selected upon creation), a permission request will be received right after. app will prompt a permission panel for user to grant permission for connection as before.
The account creation request will be stored in the current session, same behaviour as the eth/solana permission request.

This pull request fixes #6046 

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [ ] Light & dark mode
  - [ ] Different size classes (iPhone, landscape, iPad)
  - [ ] Different dynamic type sizes

## Test Plan:
1. Make sure your wallet does not have a solana account. For version with 1.44 and above, wallet creation will automatically create a solana account for you. If you need to create a new wallet with only an eth account, you would need to use an older build to do so.
2. go to a solana dapp (https://pwgoom.csb.app/) and try to connect wallet
3. a notification may appear on the top if wallet notification is not turned off in wallet settings.
4. click the notification 
5. a wallet unlock screen may appear if your wallet is locked
6. unlock your wallet
7. app will prompt an account creation screen
8. click add to create a solana account
9. app will prompt an permission screen right away
10. grant permission 
11. a. observe the wallet panel displayed the new solana account is selected, and it is connected. b. observe the dapp should display it has been connected to your wallet

1. Make sure your wallet does not have a solana account. For version with 1.44 and above, wallet creation will automatically create a solana account for you. If you need to create a new wallet with only an eth account, you would need to use an older build to do so.
2. go to a solana dapp (https://pwgoom.csb.app/) and try to connect wallet
3. a notification may appear on the top if wallet notification is not turned off in wallet settings.
4. click the notification 
5. a wallet unlock screen may appear if your wallet is locked
6. unlock your wallet
7. app will prompt an account creation screen
8. dismiss this screen by swiping down 
9. observe the wallet icon in url bar has a red badge indicates there is a pending request
10. dismiss the wallet panel 
11. click wallet icon in url bar to open wallet panel 
12. observe that the solana account creation screen shows up again
13. dismiss this screen by click cancel
14. observe the wallet icon in url bar has no red badge since user clicked `cancel` to not create a solana account
15. dismiss the wallet panel
16. click the wallet icon in url bar to open wallet panel
17. observe the create new account will not show up again

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Sources/BraveWallet/WalletProviderPermissionRequestsManager.swift [hunk 1]
```diff
@@ -86,7 +86,7 @@ public class WalletProviderPermissionRequestsManager {
   }
   
   /// Cancels all an in-flight requests without executing any decision
-  public func cancelAllPendingRequests() {
-    requests.removeAll()
+  public func cancelAllPendingRequests(coins: [BraveWallet.CoinType]) {
```
Comment (StephenHeaps) [inline Sources/BraveWallet/WalletProviderPermissionRequestsManager.swift]: nit: `public func cancelAllPendingRequests(for coins: [BraveWallet.CoinType])` (added `for`)

File: Sources/BraveWallet/WalletStrings.swift [hunk 2]
```diff
@@ -257,8 +257,8 @@ extension Strings {
       "wallet.addAccountTitle",
       tableName: "BraveWallet",
       bundle: .strings,
-      value: "Add Account",
-      comment: "The title of the add account screen"
+      value: "Add %@ Account",
+      comment: "The title of the add account screen. '%@' will be placed with a coin type title. For example, 'Add Ethereum Account'. "
```
Comment (StephenHeaps) [inline Sources/BraveWallet/WalletStrings.swift:261]: Doesn't the key need to update so we get updated translations?

File: Sources/BraveWallet/Crypto/Stores/TabDappStore.swift [hunk 3]
```diff
@@ -9,6 +9,8 @@ import Foundation
 public class TabDappStore: ObservableObject {
   /// A set of solana account addresses that are currently connected to the dapp
   @Published public var solConnectedAddresses: Set<String> = .init()
+  /// The latest dapp permission request 
+  @Published public var latestPermissionRequest: WebpagePermissionRequest?
```
Comment (StephenHeaps) [inline Sources/BraveWallet/Crypto/Stores/TabDappStore.swift]: Should this just be `pendingPermissionRequest` to match elsewhere?

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift [hunk 4]
```diff
@@ -208,6 +207,7 @@ extension Tab: BraveWalletProviderDelegate {
         self.tabDelegate?.updateURLBarWalletButton()
       })
 
+      tabDappStore.latestPermissionRequest = request
```
Comment (StephenHeaps) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift]: Should this check for nil first, or maybe better should it grab the first permission request from the permission request manager? Everywhere else we get the _first_ pending request, but I believe here could override that logic with the most recent permission request

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift [hunk 5]
```diff
@@ -271,22 +271,35 @@ extension Tab: BraveWalletProviderDelegate {
   }
   
   func showAccountCreation(_ type: BraveWallet.CoinType) {
+    let origin = getOrigin()
+    // store the account creation request
+    WalletProviderAccountCreationRequestManager.shared.beginRequest(for: origin, coinType: type) {
+      self.tabDelegate?.updateURLBarWalletButton()
```
Comment (StephenHeaps) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift]: Is weak self needed here?

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift (hunk 4 above)
Comment (nuo-xu) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift]: there is a nil check inside `WalletPanelView`.
We only care about the very first permission request right after account creation, in this case the `WalletPanelView` is not nil and will listen the change of `TabDappStore.latestPermissionRequest` and this is the first entry to capture this permission request. 
In the regular flow, when this permission request callback getting called, at that moment `WalletPanelView` has not yet been initialized. Hence, in that scenario, the change of `TabDappStore.latestPermissionRequest` won't have any impact to the wallet.
In other scenarios, which `WalletPanelView` will check the queue of stored and not yet decided permission requests and they are store in the manager class.
TLDR: `TabDappStore.latestPermissionRequest` is really just for the scenario user gets a permission request right after they created a solana account.

File: Sources/BraveWallet/Crypto/Stores/TabDappStore.swift (hunk 3 above)
Comment (nuo-xu) [reply Sources/BraveWallet/Crypto/Stores/TabDappStore.swift]: I will change it to `latestPendingPermissionRequest` just to distinguish it from `pendingRequest` from other flows.

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift (hunk 4 above)
Comment (StephenHeaps) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift]: > there is a nil check inside WalletPanelView.

Ah I was referring to checking that `TabDappStore.latestPermissionRequest` is nil before we assign a new value to it, possibly overwriting an existing permission request (even if unlikely to occur)?

Thinking further, should we even be checking `permissionRequestManager.pendingRequests(for: origin, coinType: .eth).first` in `WalletPanelView`s `onAppear` anymore? Can we use this `TabDappStore` published property exclusively? That might alleviate some of my confusion here (if it's possible) 🙂

File: Sources/BraveWallet/Panels/WalletPanelView.swift [hunk 6]
```diff
@@ -511,13 +511,32 @@ struct WalletPanelView: View {
...
+            isConnectHidden = false
+          }
+          tabDappStore.latestPermissionRequest = nil
+        }))
+      }
+    }
     .onAppear {
+      if let accountCreationRequest = WalletProviderAccountCreationRequestManager.shared.firstPendingRequest(for: origin, coinTypes: Array(WalletConstants.supportedCoinTypes)) {
+        presentWalletWithContext(.createAccount(accountCreationRequest))
+        return
```
Comment (StephenHeaps) [inline Sources/BraveWallet/Panels/WalletPanelView.swift]: I don't think we want to `return` here, as we'll miss updating pieces of `WalletPanelView` from the bottom of this `onAppear` closure

Comment (nuo-xu) [reply Sources/BraveWallet/Panels/WalletPanelView.swift]: This is due to the question we asked James at the beginning when i started integrate this. The question is if we have account creation request and permission request at the same time, which one to display first. 
The decision is to display account creation request first, that's why we check `firstPendingRequest` for account creation request first. 
In this flow,
if user creates an solana account `tabDappStore.latestPendingPermissionRequest` will take care the flow continuity to display the permission request screen.
If user ignores it, next time they are back to wallet panel, it will prompt again 
If user decide not to create, the account creation request will be removed from the queue.

Comment (StephenHeaps) [reply Sources/BraveWallet/Panels/WalletPanelView.swift]: Can't we use an else block instead of return?
```
if let accountCreationRequest = WalletProviderAccountCreationRequestManager.shared.firstPendingRequest(for: origin, coinTypes: Array(WalletConstants.supportedCoinTypes)) {
  <...>
} else if let request = WalletProviderPermissionRequestsManager.shared.pendingRequests(for: origin, coinType: .eth).first {
  <...>  
}
```

File: Sources/BraveWallet/WalletStrings.swift (hunk 2 above)
Comment (nuo-xu) [reply Sources/BraveWallet/WalletStrings.swift:261]: i dont think so. i think when the value changed it will be retranslated. Double check with the README.md under `l10n/tools`, dont think we need to update the key.
cc @iccub to confirm.

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift (hunk 4 above)
Comment (nuo-xu) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift]: we still need to check `onAppear` for user gets permission request when they have solana account (dont need to go through account creation flow)

File: Sources/BraveWallet/Panels/WalletPanelView.swift (hunk 6 above)
Comment (nuo-xu) [reply Sources/BraveWallet/Panels/WalletPanelView.swift]: sure thing. fixed at [5bfc01c](https://github.com/brave/brave-ios/pull/6075/commits/5bfc01c1dbe59d88299f4b2d3d51f6f53387901c)

File: Sources/BraveWallet/WalletProviderPermissionRequestsManager.swift (hunk 1 above)
Comment (nuo-xu) [reply Sources/BraveWallet/WalletProviderPermissionRequestsManager.swift]: fixed at [e61dd5e](https://github.com/brave/brave-ios/pull/6075/commits/e61dd5eabe27f3c7a769f1918b9d707c953083b2)

File: Sources/BraveWallet/Crypto/Stores/TabDappStore.swift (hunk 3 above)
Comment (nuo-xu) [reply Sources/BraveWallet/Crypto/Stores/TabDappStore.swift]: fixed at [e61dd5e](https://github.com/brave/brave-ios/pull/6075/commits/e61dd5eabe27f3c7a769f1918b9d707c953083b2)

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift (hunk 5 above)
Comment (nuo-xu) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift]: fixed at [e61dd5e](https://github.com/brave/brave-ios/pull/6075/commits/e61dd5eabe27f3c7a769f1918b9d707c953083b2)

File: Sources/BraveWallet/Crypto/Accounts/Add/AddAccountView.swift [hunk 7]
```diff
@@ -80,7 +84,7 @@ struct AddAccountView: View {
     .listStyle(InsetGroupedListStyle())
     .listBackgroundColor(Color(UIColor.braveGroupedBackground))
     .navigationBarTitleDisplayMode(.inline)
-    .navigationTitle(Strings.Wallet.addAccountTitle)
+    .navigationTitle(String.localizedStringWithFormat(Strings.Wallet.addAccountTitle, preSelectedCoin?.localizedTitle ?? (selectedCoin?.localizedTitle ?? BraveWallet.CoinType.eth.localizedTitle)))
```
Comment (StephenHeaps) [inline Sources/BraveWallet/Crypto/Accounts/Add/AddAccountView.swift]: This title doesn't work in the add account view coin selection screen.
Simulator Screen Shot - iPhone 13 Pro - 2022-10-18 at 22 58 07

**Review by StephenHeaps — CHANGES_REQUESTED** [review]

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift (hunk 4 above)
Comment (StephenHeaps) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+Wallet.swift]: I meant we check `tabDappStore.latestPendingPermissionRequest` in `WalletPanelView` onAppear, not removing the check there entirely. Then we are using _latest_ permission request everywhere we access permission requests.

Alternatively in this `requestPermissions(_:accounts:completion:)` function, we can check for the first available request as we are in `WalletPanelView` onAppear now, wouldn’t it provide the same request in the flow you describe?
```
if tabDappStore.pendingPermissionRequest == nil {
  tabDappStore.pendingPermissionRequest = permissionRequestManager.firstPendingRequest(for: origin, coinTypes: [.eth, .sol])
}
```

I just prefer consistency, there doesn’t seem to be a reason for these to be first request in one spot and last request in another?

File: Sources/BraveWallet/WalletStrings.swift (hunk 2 above)
Comment (StephenHeaps) [reply Sources/BraveWallet/WalletStrings.swift:261]: Resolved with this change request: https://github.com/brave/brave-ios/pull/6075#discussion_r998897468
This key remains now, and a new localization is added for new title.

**Review by StephenHeaps — APPROVED** [review]
